model.py

```diff
@@ -198,6 +208,40 @@ def forward(self, x: Tensor) -> Tensor:
         return self.w2(F.silu(self.w1(x)) * self.w3(x))
 
 
+class ConditionalFeedForward(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.w1 = nn.Parameter(torch.empty(config.num_experts, config.intermediate_size, config.dim))
+        self.w2 = nn.Parameter(torch.empty(config.num_experts, config.intermediate_size, config.dim))
+        self.w3 = nn.Parameter(torch.empty(config.num_experts, config.intermediate_size, config.dim))
+
+    def forward(self, x: Tensor, expert_indices: Tensor) -> Tensor:
+        w1_weights = self.w1[expert_indices].transpose(-1, -2) # [T, A, D, D]
+        w3_weights = self.w3[expert_indices].transpose(-1, -2) # [T, A, D, D]
+        w2_weights = self.w2[expert_indices]  # [T, A, D, D]
+        x1 = F.silu(torch.einsum('ti,taio -> tao', x, w1_weights))
+        x3 = torch.einsum('ti, taio -> tao', x, w3_weights)
+        expert_outs =  torch.einsum('tao, taoi -> tai', (x1 * x3), w2_weights)
+        return expert_outs
+
+
+class MOEFeedForward(nn.Module):
+    def __init__(self, config) -> None:
+        super().__init__()
+        self.gate = nn.Linear(config.dim, config.num_experts, bias=False)
+        self.cond_ffn = ConditionalFeedForward(config)
+        self.dim = config.dim
+    def forward(self, x: Tensor) -> Tensor:
+        x = x.view(-1, self.dim)
+        # T = num_tokens, E = num_experts, D = hidden dim, A = activated experts
+        # x: [T, D]
+        scores = self.gate(x) # [T, E]
+        expert_weights, expert_indices = torch.topk(scores, 2, dim=-1) # [T, A], [T, A]
```

Comment:
Fwiw, I think this is a little different than the [reference implementation](https://github.com/mistralai/vllm-release/blob/f47ae04cce40b25881eedb31dbb66f6a17674408/vllm/model_executor/models/mixtral.py#L344) from Mistral?

They do softmax -> topk -> norm. This looks like it does topk -> softmax.

Reply:
You are right, I just updated it. Thanks for catching this!

Reply:
Happy to help :)